client/mm/libxc/coinbase.go

```diff
@@ -0,0 +1,1679 @@
+	block, _ := pem.Decode([]byte(apiPrivateKey))
+	if block == nil {
+		return "", fmt.Errorf("jwt: Could not decode private key")
+	}
+
+	key, err := x509.ParseECPrivateKey(block.Bytes)
+	if err != nil {
+		return "", fmt.Errorf("jwt: %w", err)
+	}
+
+	sig, err := jose.NewSigner(
+		jose.SigningKey{Algorithm: jose.ES256, Key: key},
+		(&jose.SignerOptions{NonceSource: nonceSource{}}).WithType("JWT").WithHeader("kid", apiName),
+	)
+	if err != nil {
+		return "", fmt.Errorf("jwt: %w", err)
+	}
+
+	cl := &APIKeyClaims{
+		Claims: &jwt.Claims{
+			Subject:   apiName,
+			Issuer:    "coinbase-cloud",
+			NotBefore: jwt.NewNumericDate(time.Now()),
+			Expiry:    jwt.NewNumericDate(time.Now().Add(time.Minute * 2)),
+		},
+		URI: uri,
+	}
+	jwtString, err := jwt.Signed(sig).Claims(cl).CompactSerialize()
+	if err != nil {
+		return "", fmt.Errorf("jwt: %w", err)
+	}
+	return jwtString, nil
+}
+
+func (c *coinbase) sign(preimage string) (string, error) {
+	key, err := base64.StdEncoding.DecodeString(c.apiPrivateKey)
+	if err != nil {
+		return "", err
+	}
+
+	signature := hmac.New(sha256.New, key)
+	_, err = signature.Write([]byte(preimage))
+	if err != nil {
+		return "", err
+	}
+
+	return base64.StdEncoding.EncodeToString(signature.Sum(nil)), nil
+}
+
+func (c *coinbase) prepareRequest(ctx context.Context, method, endpoint string, queryParams url.Values, bodyParams interface{}) (_ *http.Request, _ context.CancelFunc, err error) {
+	var body []byte
+	if bodyParams != nil {
+		body, err = json.Marshal(bodyParams)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error marshaling request: %w", err)
+		}
+	}
+
+	fullUrl := fmt.Sprintf("https://%s%s", c.basePath, endpoint)
+	if queryParams != nil {
+		fullUrl += "?" + queryParams.Encode()
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second*30)
+	defer func() {
+		if err != nil {
+			cancel()
+		}
+	}()
+
+	req, err := http.NewRequestWithContext(ctx, method, fullUrl, bytes.NewReader(body))
+	if err != nil {
+		return nil, nil, fmt.Errorf("error generating http request: %w", err)
+	}
+
+	uri := fmt.Sprintf("%s %s%s", method, c.basePath, endpoint)
+	jwt, err := buildJWT(c.apiName, c.apiPrivateKey, uri)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error building JWT: %w", err)
+	}
+	req.Header.Set("Authorization", "Bearer "+jwt)
+	req.Header.Set("CB-VERSION", "2024-05-27")
+
+	return req, cancel, nil
+}
+
+func (c *coinbase) updateAccounts(ctx context.Context) error {
+	accounts := make(map[string]*cbtypes.Account)
+
+	var cursor string
+	hasNext := true
+	// The accounts are paginated. If there are more than 250 accounts, we need to
+	// fetch additional.
+	for i := 0; i < 20 && hasNext; i++ {
+		v := make(url.Values)
+		if cursor != "" {
+			v.Add("cursor", cursor)
+		}
+		v.Add("limit", "250")
+
+		var res cbtypes.AccountsResult
+		if err := c.request(ctx, http.MethodGet, "api/v3/brokerage/accounts", v, nil, &res); err != nil {
```

Comment:
Hmm. I also see docs at https://docs.cdp.coinbase.com/exchange/reference/exchangerestapi_getaccount that are a little different.

Reply:
You're looking at the Coinbase Exchange API, which is for institutions. We're using the Advanced Trade API, for individuals.

https://docs.cdp.coinbase.com/advanced-trade/reference/retailbrokerageapi_getaccounts